### PR TITLE
Proposal for RPC tokens

### PR DESCRIPTION
**Pull request by yichenshen**
This is the proposal for using RPC tokens to isolate rings from improperly setup instances.

The PR is large, but github doesn't let me stack changes. I suggest looking through the individual commits which I've split out.

The gist is this:

1. We identify the Raft related RPCs are that done automatically by kuduraft, such as 
  -  UpdateConsensus
  - RequestConsensusVote
  - RunLeaderElection
2. We then stamp each request and each response with a field: raft_rpc_token
3. In ConsensusService, in enforce mode, we reject outright all such requests without a matching token
4. In PeerProxy we hijack the response callback and replace responses with errors if the response token is wrong

This way, we intercept all RPC communication without a proper token before it hit any part of raft_consensus. Instances with a bad token will essentially act like they've been cut off from the ring since all their RPCs are discarded.

Couple of design explanations:

## Rejecting responses

Initially I only rejected requests. However, there are edge cases. A bad instance joining a ring with a higher term and is itself not enforcing for request tokens will happily accept RPCs from the leader and respond. Based on the Raft protocol the higher term in the response will cause the current leader to step down. If the bad instance keeps doing this, even with request tokens a bad instance can still destabilize the ring. This is why we need to discard responses from the bad instance too.

## Enforcement
Currently we have 2 levels, on or off. When it's ON will actively generate errors to interrupt the RPC flow. When it's OFF we currently just log a WARNING log and pretend nothing has happened. Either way tokens are compared.

This needs to be improved to bump counters, but this PR is huge as it is so that'll be a next step.

## Token storage

Tokens are stored in persistent_vars, which will survive a restart. They can be set via public APIs of RaftConsensus. persistent_var is not thread safe.

Tokens are used in the critical Raft replication path, so I designed them to be lock free on reads by using atomic ops on a shared ptr. Writes are synchronized by the various locks in RaftConsensus and Peers but we 

We don't allow tokens to change when we're enforcing. It's technically possible for one or two rpcs to miss the new token if we set a new token and turn on enforcement immediately (which is not really an intended use case anyway, we should set the token, do verifications, then enforce). However, even in such a case the next RPC will know of the new token since it's read from the ptr every time.

### CODE REVIEW

**Comment by li-chi** [comment]
topology_config automatically persists on disk. Since ring_uuid is already part of topology_config, why not just use it?

**Comment by yichenshen** [comment]
> topology_config automatically persists on disk. Since ring_uuid is already part of topology_config, why not just use it?

topology_config is used for bootstrapping, and I believe it's only persisted on init. We need something that persists on the fly. Persistent_vars already gives us that for free.

Another point is that if we do it in topology_config, it'll be passed to kuduraft as a server_opt. server_opt will eventually be copied to actual object properties when we bootstrap. Most likely this will live in RaftConsensus. But we need to reference this in ConsensusQueue as well, and if it's changed, we'll have to thread it into ConsensusQueue.

Persistent vars is managed by a persistent_vars_manager and acts like a singleton. This makes it easier to set and read from multiple places without having to do extra threading.

File: src/kudu/tserver/consensus_service.cc [hunk 1]
```diff
@@ -430,6 +491,11 @@ void ConsensusServiceImpl::RunLeaderElection(const RunLeaderElectionRequestPB* r
   shared_ptr<RaftConsensus> consensus;
   if (!GetConsensusOrRespond(tablet_manager_, resp, context, &consensus)) return;
 
+  if (!CheckRaftRpcTokenOrRespond("RunLeaderElection", req, resp, context,
```
Comment (li-chi) [inline src/kudu/tserver/consensus_service.cc:493]: Why does not this one have "resp->set_raft_rpc_token(*std::move(ownToken));"?

File: src/kudu/consensus/consensus_peers.cc [hunk 2]
```diff
@@ -616,6 +619,47 @@ void PeerProxyPool::Clear() {
   peer_proxy_map_.clear();
 }
 
+template <class RespType>
+void CheckAndEnforceResponseToken(const std::string& method_name,
+                                  RespType* response,
+                                  boost::optional<std::string> rpc_token,
+                                  bool enforce) {
```
Comment (li-chi) [inline src/kudu/consensus/consensus_peers.cc]: Why is CheckRaftRpcTokenOrRespond() not enough?

File: src/kudu/consensus/consensus_peers.cc [hunk 3]
```diff
@@ -629,7 +673,20 @@ void RpcPeerProxy::UpdateAsync(const ConsensusRequestPB* request,
                                rpc::RpcController* controller,
                                const rpc::ResponseCallback& callback) {
   controller->set_timeout(MonoDelta::FromMilliseconds(FLAGS_consensus_rpc_timeout_ms));
-  consensus_proxy_->UpdateConsensusAsync(*request, response, controller, callback);
+
+  boost::optional<std::string> rpc_token = request->has_raft_rpc_token()
+                                               ? request->raft_rpc_token()
+                                               : boost::optional<std::string>();
+  bool enforce = controller->enforce_rpc_token();
```
Comment (li-chi) [inline src/kudu/consensus/consensus_peers.cc]: What is the reason for needing extra copy in rpc_controller, and not using ShouldEnforceRaftRpcToken() in raft_consensus or in consensus_queue? Is it because consensus_peer does not have reference to them?

Having three copies of the same value makes update harder.

File: src/kudu/consensus/consensus_peers.cc (hunk 2 above)
Comment (yichenshen) [reply src/kudu/consensus/consensus_peers.cc]: This class is on the outgoing request path and we're checking that the **response** returned to us has the right token.

CheckRaftRpcTokenOrRespond() is in the incoming request path where we check the **request** sent to us has the right token.

We need checks on both sides because both requests and responses trigger side effects. Responses for example can trigger term advancements and we need to reject them for bad instances.

File: src/kudu/consensus/consensus_peers.cc (hunk 3 above)
Comment (yichenshen) [reply src/kudu/consensus/consensus_peers.cc]: consensus_peers is a separate module from raft_consensus and consensus_queue. It shouldn't know about any of those as part of the abstraction contract.

rpc_controller is initialized per request, so there's no need for any action when the value is updated. The next request will simply pick it up.

File: src/kudu/tserver/consensus_service.cc (hunk 1 above)
Comment (yichenshen) [reply src/kudu/tserver/consensus_service.cc:493]: Because we do nothing on the response. We can technically add it for completeness, but today if you look at the call in consensus_peers, you'll see that we don't even have a callback to handle the response we receive, so even if we set it, no one will read it.

File: src/kudu/tserver/consensus_service.cc [hunk 4]
```diff
@@ -286,6 +325,17 @@ void ConsensusServiceImpl::UpdateConsensus(const ConsensusRequestPB* req,
   shared_ptr<RaftConsensus> consensus;
   if (!GetConsensusOrRespond(tablet_manager_, resp, context, &consensus)) return;
 
+  if (auto ownToken = consensus->RaftRpcToken()) {
```
Comment (li-chi) [inline src/kudu/tserver/consensus_service.cc]: nit: Move below CheckRaftRpcTokenOrRespond(), as we do not even respond, let long need token, if the sender has wrong token.

**Review by li-chi — APPROVED** [review]
Looks good to me. Having the ability to get some statistic and confidence before enforcing strict mode would be good.

File: src/kudu/consensus/consensus_peers.cc (hunk 2 above)
Comment (anirbanr-fb) [reply src/kudu/consensus/consensus_peers.cc]: There are multiple kinds of error messages on requests. We don't have to send a term back on a bad request. We can just drop it. In that case response path should not have to do anything?

**Review by anirbanr-fb — COMMENTED** [review]
General comment as previous diff. I don't think the response path is required.
Its like if someone is trying to DDOS you, sending him any information about you
back might not be a good idea. Even if we are sending a response back, giving
a response like 503 which essentially gives the UUID but does not put a term is the
right thing to do, since we don't want the requester to step down.

In the case of sending a UUID back on response, it can help debugging the issue on requester side,
but we need to eliminate the risk of step down. We can thereby reduce response handling code as well.

**Comment by yichenshen** [comment]
But on the destination side this is already what I'm doing. If the destination finds a mismatch it already sends back RING_TOKEN_MISMATCH as an RPC error and no rpc information. 

The response checking is for another real scenario I saw during testing: 

I started a ring, set the token and made one node bad. I turned on enforcement for everyone else.

The leader reached out to the bad node. The bad node didn't care about the token (because it's not enforcing), and just gave a reply with a higher term. The leader gets this higher term and immediately stepped down. In other words, even though the leader is enforcing on requests, it didn't reject a bad response. And since leader step down is an action we take on receiving a response, just checking when we get a request is insufficient. So a bad instance, even if the leader enforces requests checks, can still destablize the ring.

So the case I'm guarding here is when a destination (in the config) acts in bad faith and did not enforce the contract. I'm adding the requirement that followers present their own token on a response, so that if the follower is misconfigured, the leader will not care about their responses.

File: src/kudu/consensus/metadata.proto [hunk 5]
```diff
@@ -160,6 +160,11 @@ message TopologyConfigPB {
   // Caution: This should NOT be modified by automation when
   // servers are being removed for maintenance.
   map<string, int32> voter_distribution = 8;
+
+  // The initial raft rpc token used to prove we're part of the ring.
+  // This will be populated if we don't already have a rpc token from
+  // persistent vars
+  optional string initial_raft_rpc_token = 9;
```
Comment (anirbanr-fb) [inline src/kudu/consensus/metadata.proto:167]: nit: is the word initial required?

Comment (yichenshen) [reply src/kudu/consensus/metadata.proto:167]: Yes, this is like the bootstrap topology config, only works the first time. I want to make that clear.

File: src/kudu/consensus/raft_consensus.h [hunk 6]
```diff
@@ -250,6 +251,12 @@ class RaftConsensus : public std::enable_shared_from_this<RaftConsensus>,
   // Check if starting elections is allowed
   bool IsStartElectionAllowed() const;
 
+  // Sets a RPC token to be sent with Raft RPCs to prove we're in a certain ring
+  void SetRaftRpcToken(boost::optional<std::string> token);
+
+  // Returns the rpc token
+  std::shared_ptr<const std::string> RaftRpcToken() const;
```
Comment (anirbanr-fb) [inline src/kudu/consensus/raft_consensus.h]: nit: in most get functions at this class we are using an explicit Get.
e.g. GetLeaderUuid/GetProxyTopology

Comment (anirbanr-fb) [reply src/kudu/consensus/raft_consensus.h]: Regarding your point about when a leader is enforcing and the destination is not enforcing. That is a partial
state of enforcing and a temporary state during rollouts. To borrow from SSL both sides have to be enforcing(e.g. checking certificates)  for the system overall to function as expected. So to me the partial state was hit in a
test, and is no different the behavior of the LEADER just before it started enforcing.

We can discuss this more as I haven't reviewed the entire PR, but this might be a bit overdesign/overkill (but still agree that mostly boiler plate)

Comment (yichenshen) [reply src/kudu/consensus/raft_consensus.h]: Yes, if we eventually move to a model where enforcement is mandatory and unconfigurable then there's no possibility.

But I'm not sure if we'll ever be in that state. We want knobs to be able to disable it if needed (we also need to disable it if we want to change the token), and as long a the knobs are there we can misconfigure it. We can expect the misconfigured instance to not work, but I think it's reasonable to set a bar where a misconfigured instance can't bring the rest of the ring down.

File: src/kudu/consensus/raft_consensus.cc [hunk 7]
```diff
@@ -3503,6 +3508,16 @@ bool RaftConsensus::IsStartElectionAllowed() const {
   return persistent_vars_->is_start_election_allowed();
 }
 
+void RaftConsensus::SetRaftRpcToken(boost::optional<std::string> token) {
+  LockGuard guard(lock_);
```
Comment (anirbanr-fb) [inline src/kudu/consensus/raft_consensus.cc:3523]: we should log here, because it is a significant event from kuduraft perspective

File: src/kudu/consensus/raft_consensus.cc [hunk 8]
```diff
@@ -309,6 +309,11 @@ Status RaftConsensus::Init() {
 
   RETURN_NOT_OK(persistent_vars_manager_->LoadPersistentVars(options_.tablet_id, &persistent_vars_));
 
+  if (!persistent_vars_->raft_rpc_token()) {
+    persistent_vars_->set_raft_rpc_token(options_.initial_raft_rpc_token);
+    CHECK_OK(persistent_vars_->Flush());
```
Comment (anirbanr-fb) [inline src/kudu/consensus/raft_consensus.cc:318]: are we validating the initial_raft_rpc_token. At some point in time, new rings should enforce this (i.e. should not be empty) ?

**Review by anirbanr-fb — COMMENTED** [review]
this particular commit looks good to me

File: src/kudu/consensus/raft_consensus.cc (hunk 8 above)
Comment (yichenshen) [reply src/kudu/consensus/raft_consensus.cc:318]: At some point we can enforce that after this if block we have a token I think. That would make sense.

initial_raft_rpc_token is from topology config, and it might always be empty for the instances we currently have (which will populate directly from persistent_vars after we do a rollout override). We'll have to either wait for them to be replaced or ninja land some kind of json change into topology config

**Review by anirbanr-fb — COMMENTED** [review]
The commit "Get relevant RPCs to stamp the RPC tokens" looks good to me.

File: src/kudu/consensus/consensus_queue.cc [hunk 9]
```diff
@@ -175,6 +175,7 @@ PeerMessageQueue::Metrics::Metrics(const scoped_refptr<MetricEntity>& metric_ent
 PeerMessageQueue::PeerMessageQueue(const scoped_refptr<MetricEntity>& metric_entity,
                                    scoped_refptr<log::Log> log,
                                    scoped_refptr<ITimeManager> time_manager,
+                                   const scoped_refptr<PersistentVarsManager>& persistent_vars_manager,
```
Comment (anirbanr-fb) [inline src/kudu/consensus/consensus_queue.cc:178]: why are we funneling persistent vars manager into consensus queue, instead of funneling persistent vars_?
That way we don't have to LoadPersistentVars again.

File: src/kudu/consensus/consensus.proto [hunk 10]
```diff
@@ -373,6 +376,10 @@ message VoteRequestPB {
   // Identifies the tablet configuration a the vote is being requested for.
   required string tablet_id = 1;
 
+  // A token stamped to the request to prove to the remote host that we're part
+  // of a ring
+  optional string raft_rpc_token = 9;
```
Comment (anirbanr-fb) [inline src/kudu/consensus/consensus.proto]: we should make this the last item in the struct for readability.

File: src/kudu/consensus/consensus.proto [hunk 11]
```diff
@@ -457,6 +464,10 @@ message ConsensusRequestPB {
 
   required string tablet_id = 1;
 
+  // A token stamped to the request to prove to the remote host that we're part
+  // of a ring
```
Comment (anirbanr-fb) [inline src/kudu/consensus/consensus.proto]: same here - last item in list to be consistent with number

File: src/kudu/consensus/consensus_peers.cc [hunk 12]
```diff
@@ -292,9 +292,10 @@ void Peer::SendNextRequest(bool even_if_queue_empty, bool from_heartbeater) {
   string next_hop_uuid;
   int64_t commit_index_before = request_.has_committed_index() ?
       request_.committed_index() : kMinimumOpIdIndex;
-  Status s = queue_->RequestForPeer(peer_pb_.permanent_uuid(), read_ops, &request_,
-                                    &replicate_msg_refs_, &needs_tablet_copy,
-                                    &next_hop_uuid);
+  bool enforce_response_token;
+  Status s = queue_->RequestForPeer(
+      peer_pb_.permanent_uuid(), read_ops, &request_, &replicate_msg_refs_,
+      &needs_tablet_copy, &next_hop_uuid, &enforce_response_token);
```
Comment (anirbanr-fb) [inline src/kudu/consensus/consensus_peers.cc]: why is enforce_response_token a return output here. At the time of start of request, i.e. before RequestForPeer we should know whether we want to enforce or not?

File: src/kudu/consensus/consensus_queue.cc [hunk 13]
```diff
@@ -747,6 +755,8 @@ Status PeerMessageQueue::RequestForPeer(const string& uuid,
         *next_hop_uuid = uuid;
       }
     }
+
+    *enforce_response_rpc_token = enforce_response_rpc_token_;
```
Comment (anirbanr-fb) [inline src/kudu/consensus/consensus_queue.cc]: why is this required. We seem to be just passing a pointer and getting back the local variable.

File: src/kudu/consensus/leader_election.h [hunk 14]
```diff
@@ -434,6 +434,7 @@ class LeaderElection : public RefCountedThreadSafe<LeaderElection> {
                  VoteRequestPB request,
                  gscoped_ptr<VoteCounter> vote_counter,
                  MonoDelta timeout,
+                 bool enforce_rpc_token,
```
Comment (anirbanr-fb) [inline src/kudu/consensus/leader_election.h]: why is leader election stating to have concepts like enforce_rpc_token in its interface.
We seem to be leaking very low level concepts into a higher level interface.

File: src/kudu/consensus/consensus_peers.cc (hunk 3 above)
Comment (anirbanr-fb) [reply src/kudu/consensus/consensus_peers.cc]: I agree with the general sentiment here brought up by Chi. We are threading enforce_rpc_token into all parts of the system which don't need to have any threading required. This is not a knob that is going to be required after rollout phase and too low level for threading overhead. We should try to use a global, or a single call somewhere to whether to enforce or not.

File: src/kudu/consensus/consensus_queue.cc (hunk 9 above)
Comment (yichenshen) [reply src/kudu/consensus/consensus_queue.cc:178]: We don't load the persistent vars (from protobuf) here either. persistent_vars_manager holds a cached map of all loaded persistent vars. It's loaded once and subsequent calls just returns a pointer from the map.

It's essentially acting like a per tablet singleton. Ofc we only have one tablet so how we pass it is a cosmetic difference. I passed the manager mainly to keep things consistent with other places that uses persistent vars.

File: src/kudu/consensus/consensus_peers.cc (hunk 12 above)
Comment (yichenshen) [reply src/kudu/consensus/consensus_peers.cc]: We actually don't, because the peer object is created once per known peer and it's state is not tied to any request.

This is why we call back into the queue to ask for a request to be built for the peer. Whether we enforce the token is also a request state, which is why it has to be returned here.

File: src/kudu/consensus/consensus_queue.cc (hunk 13 above)
Comment (yichenshen) [reply src/kudu/consensus/consensus_queue.cc]: Yes, because one is a queue state and one is a request state. We copy the queue state for this particular request, which is the job of this function. We create a request for Peer using the current queue state.

The queue can move on after, but the request params are snapshotted.

File: src/kudu/consensus/leader_election.h (hunk 14 above)
Comment (yichenshen) [reply src/kudu/consensus/leader_election.h]: This has to do with who is initiating the rpc request. Currently the way it's done is that it's set by the initiator. Leader election is at the same level as RaftConsensus; they both are the initiator of the RPC, which is why this is here.

I was struggling a bit with this mostly because the actual best place to do it it probably in ConsensusService, but it was hard to get the option into there.

That can be re-examined.

File: src/kudu/tserver/consensus_service.cc [hunk 15]
```diff
@@ -214,6 +214,44 @@ bool GetConsensusOrRespond(TSTabletManager* tablet_manager,
...
+  }
+
+  auto error_message = Substitute(
+      "Raft RPC token mismatch. Receiver token: $0. Request token: $1",
+      ownToken ? *ownToken : "<null>",
+      req->has_raft_rpc_token() ? req->raft_rpc_token() : "<null>");
+
+  if (!consensus.ShouldEnforceRaftRpcToken()) {
+    // Mismatch but don't enforce
+    KLOG_EVERY_N(WARNING, 5000)
```
Comment (anirbanr-fb) [inline src/kudu/tserver/consensus_service.cc]: would time based be less chatty?

Comment (yichenshen) [reply src/kudu/tserver/consensus_service.cc]: Probably similar because our heartbeats are regular enough, but I like the idea because it'll be more predictable.
